timemachine/cpp/src/kernels/k_exchange.cu

```diff
@@ -888,4 +846,102 @@ template void __global__ k_accept_first_valid_move<double>(
     int *__restrict__ accepted_sample          // [1]
 );
 
+template <typename RealType>
+void __global__ k_accept_first_valid_move_targeted(
+    const int total_proposals,
+    const int num_target_mols,
+    const int batch_size,
+    const RealType inner_volume,
+    const int *__restrict__ targeting_inner_volume, // [batch_size]
+    const int *__restrict__ inner_count,            // [1]
+    const RealType *__restrict__ box_vol,           // [1]
+    const int *__restrict__ noise_offset,           // [1]
+    const int *__restrict__ samples,                // [batch_size]
+    const RealType *__restrict__ before_max,        // [batch_size]
+    const RealType *__restrict__ before_log_sum,    // [batch_size]
+    const RealType *__restrict__ after_max,         // [batch_size]
+    const RealType *__restrict__ after_log_sum,     // [batch_size]
+    const RealType *__restrict__ rand,              // [total_proposals]
+    int *__restrict__ accepted_sample               // [1]
+) {
+    volatile __shared__ int selected_idx;
+    assert(blockIdx.x == 0);
+    if (threadIdx.x == 0) {
+        selected_idx = batch_size;
+    }
+    __syncthreads();
+    int idx = blockIdx.x * blockDim.x + threadIdx.x;
+    const int rand_offset = noise_offset[0];
+    const int local_inner_count = inner_count[0];
+    const RealType outer_vol = box_vol[0] - inner_volume;
+
+    while (idx < batch_size) {
+        if (rand_offset + idx >= total_proposals) {
+            break;
+        }
+
+        const RealType raw_log_acceptance = compute_raw_log_probability_targeted<RealType>(
+            targeting_inner_volume[idx],
+            inner_volume,
+            outer_vol,
+            local_inner_count,
+            num_target_mols,
+            before_max + idx,
+            before_log_sum + idx,
+            after_max + idx,
+            after_log_sum + idx);
+
+        const RealType log_acceptance_prob = min(raw_log_acceptance, static_cast<RealType>(0.0));
+        const bool accepted = rand[rand_offset + idx] < exp(log_acceptance_prob);
+        if (accepted) {
+            atomicMin(&selected_idx, idx);
+            // Idx is increasing so the first accepted is the min value the thread can accept
+            // making it safe to break early.
+            break;
```

Comment:
For performance reasons, reduces the potential number of `atomicMin` calls on `selected_idx` which would be pointless since any subsequent `idx` values would be greater.